core/src/server/access_control/mod.rs

```diff
@@ -0,0 +1,167 @@
+//! Access control based on HTTP headers
+
+pub mod cors;
+pub mod host;
+mod matcher;
+
+pub use cors::{AllowHeaders, AllowOrigin, Origin};
+pub use host::{AllowHosts, Host};
+
+use crate::Error;
+
+use self::cors::get_cors_allow_origin;
+
+/// Define access on control on HTTP layer.
+#[derive(Clone, Debug)]
+pub struct AccessControl {
+	allowed_hosts: AllowHosts,
+	allowed_origins: Option<Vec<AllowOrigin>>,
+	allowed_headers: AllowHeaders,
+}
+
+impl AccessControl {
+	/// Validate incoming request by HTTP HOST
+	///
+	/// `host` is the return value from the `host header`
+	pub fn verify_host(&self, host: &str) -> Result<(), Error> {
+		self.allowed_hosts.verify(host)
+	}
+
+	/// Validate incoming request by CORS origin value
+	///
+	/// `host` is the return value from the `host header`
+	/// `origin` is the value from the `origin header`.
+	pub fn verify_origin(&self, origin: Option<&str>, host: &str) -> Result<(), Error> {
+		if let cors::AllowCors::Invalid = get_cors_allow_origin(origin, &self.allowed_origins, Some(host)) {
+			Err(Error::HttpHeaderRejected("origin", origin.unwrap_or("<missing>").into()))
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Validate incoming request by CORS header
+	///
+	/// `header_names` is the headers names in the request.
+	/// `cors_headers`: is the header names from `access-control-request-headers header`.
+	pub fn verify_headers<T, I, II>(&self, header_names: I, cors_headers: II) -> Result<(), Error>
+	where
+		T: AsRef<str>,
+		I: Iterator<Item = T>,
+		II: Iterator<Item = T>,
+	{
+		let header = cors::get_cors_allow_headers(header_names, cors_headers, &self.allowed_headers, |name| name);
+
+		if let cors::AllowCors::Invalid = header {
+			Err(Error::HttpHeaderRejected("access-control-request-headers", "<too long to be logged>".into()))
```

Comment:
Why "too long to be logged"?

Reply:
I think I was just lazy because the actual values were moved into the function above.

Reply:
Hrm yeah, I mulled over this a bit and I can see that it'd be a bit of a pain!